### ISSUE TITLE
frame lost during animation

### ISSUE THREAD

**Issue by Alex1989Wang** [airbnb/lottie-ios#288]
Dear Lottie Coder, 

I am a developer from China. I am so grateful that Lottie was created. While using this wonderful framework, I found that when the difference between the inFrame and outFrame of a LOTLayer is small (in my JSON file, the difference is about one frame) the contents of the layer never shows. 

We are designing a "fireworks" animation effect. I downloaded your Github Lottie project. I included the JSON file and all needed image assets in this project. Scrolling down to the bottom of the example animations, you will find the one named "gift_fireworks". 

![gift_fireworks](https://user-images.githubusercontent.com/4620388/27734611-2122fade-5dce-11e7-9373-7c9ec41c760f.png)

Here is what I did to prevent the animation from losing frames. 

In the source file named "LOTLayer.m", within the method implementation :
```objc
- (void)_mapFromJSON:(NSDictionary *)jsonDictionary
      withCompBounds:(CGRect)compBounds
       withFramerate:(NSNumber *)framerate
     withAssetGroup:(LOTAssetGroup *)assetGroup
````

I made the following change after reading inFrame and outFrame values from the JSON file.
```objc
    if (_outFrame.integerValue - _inFrame.integerValue <= 1) {
        NSInteger offfset = 50;
        _outFrame = [NSNumber numberWithInteger:_inFrame.integerValue + offfset];
    }
```

![gift_firework_code_change](https://user-images.githubusercontent.com/4620388/27734616-2b5bb130-5dce-11e7-9647-245bec9a3691.png)

Frames won't lose anymore, but the animation effect is changed because some frames are lingering longer than they should be. And the more interesting thing is that if you set the above offset value to be 15, lost frames in the beginning stage of the animation will be shown but not lost frames in the middle or the end stage. Setting it to be 30, lost frames in the beginning and middle stage of the animation will be shown, but not those from the end stage. 

I hope you could help solve this problem. I have read the source code for two consecutive days and can't figure out why. 

The attachment is the modified project. 
https://drive.google.com/file/d/0B4zUyq9DNJp7RmZiOF96NWhDRmM/view

**Comment by buba447** [comment]
Hello! Yes this is an issue we are tracking. 
The problem comes froma small inconsistency with how CAAnimation syncs animations. If animations arent grouped together they are not guaranteed to sync exactly and can be off by a little bit. Ive also been trying to debug it for a few days and i know that tweaking the values here can fix the problem but im not confident enough in it to actually merge it into the master repo.

You can change it locally though and see if it helps. We have a new system for in out animations coming in version 2.0 of the freamework

https://github.com/airbnb/lottie-ios/blob/master/lottie-ios/Classes/Models/LOTLayer.m#L172-L191

**Comment by Alex1989Wang** [comment]
The in-out animation is added to LOTLayerView, but all the other animations are added to a container layer on LOTLayerView (in fact a layer). If the frame-lose problem is due to the out of sync between the in-out CAKeyFrameAnimation and all the other grouped animations, is it possible to get ride of the container layer and put all animations in a single grouped animation?

**Comment by Alex1989Wang** [comment]
Besides, may I ask when do you plan to release the 2.0.0 version? Our project is kind of in a rush right now?

**Comment by panmingyang2009** [comment]
I'm facing with the same problem. In my case, the interval of in out animation is 2. 

My in out animation is normal if animation materials are dragged into project.
However, if materials are downloaded to document in runtime, the animation goes wrong.

When the error occur, the animation lose some frames regularly. Sometimes 1/4 frames of each animation were lost, sometimes 3/4, sometimes all of them.

Increasing duration of each animation like Alex did can avoid the problem temporarily, but not a final solution for the problem.

What confuse me most is the different behavior between downloading to document and dragging into project. How it happened?

**Comment by buba447** [comment]
Should be fixed in Lottie 2.0! If you still have a problem please create a new issue!
